File: 0530-minimum-absolute-difference-in-bst/0530-minimum-absolute-difference-in-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    prev = 2 ** 32 - 1
    min_val = prev
    def helper(self, root: Optional[TreeNode]) -> None:
        if not root:
            return
        self.helper(root.left)
        self.min_val = min(self.min_val, abs(root.val - self.prev))
        self.prev = root.val
        self.helper(root.right)

    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:    
        self.helper(root)
        return self.min_val

File: 0530-minimum-absolute-difference-in-bst/0530-minimum-absolute-difference-in-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* root, int& min, int& prev) {
        if (!root) {
            return;
        }
        helper(root->left, min, prev);
        min = std::min(min, abs(prev - root->val));
        prev = root->val;
        helper(root->right, min, prev);
    }
    int getMinimumDifference(TreeNode* root) {
        int min = INT_MAX;
        int prev = INT_MAX;
        helper(root, min, prev);
        return min;
    }
};